In [1]:
import pandas as pd
data = pd.read_csv('FlightDelays.csv')
data.head()

CRS_DEP_TIME CARRIER  DEP_TIME DEST  DISTANCE     FL_DATE  FL_NUM ORIGIN  \
0          1455      OH      1455  JFK       184  01/01/2004    5935    BWI   
1          1640      DH      1640  JFK       213  01/01/2004    6155    DCA   
2          1245      DH      1245  LGA       229  01/01/2004    7208    IAD   
3          1715      DH      1709  LGA       229  01/01/2004    7215    IAD   
4          1039      DH      1035  LGA       229  01/01/2004    7792    IAD   

   Weather  DAY_WEEK  DAY_OF_MONTH TAIL_NUM Flight Status  
0        0         4             1   N940CA        ontime  
1        0         4             1   N405FJ        ontime  
2        0         4             1   N695BR        ontime  
3        0         4             1   N662BR        ontime  
4        0         4             1   N698BR        ontime

In [2]:
data.shape

(2201, 13)

In [4]:
import pandas as pd
data = pd.read_csv('FlightDelays.csv', header = [0])
data.head()

#header=0 -> columns in the first row in the file and if you don’t you will have to specify header=None
#0 is default

CRS_DEP_TIME CARRIER  DEP_TIME DEST  DISTANCE     FL_DATE  FL_NUM ORIGIN  \
0          1455      OH      1455  JFK       184  01/01/2004    5935    BWI   
1          1640      DH      1640  JFK       213  01/01/2004    6155    DCA   
2          1245      DH      1245  LGA       229  01/01/2004    7208    IAD   
3          1715      DH      1709  LGA       229  01/01/2004    7215    IAD   
4          1039      DH      1035  LGA       229  01/01/2004    7792    IAD   

   Weather  DAY_WEEK  DAY_OF_MONTH TAIL_NUM Flight Status  
0        0         4             1   N940CA        ontime  
1        0         4             1   N405FJ        ontime  
2        0         4             1   N695BR        ontime  
3        0         4             1   N662BR        ontime  
4        0         4             1   N698BR        ontime

In [5]:
data.shape

(2201, 13)

Summary

In [6]:
data.dtypes

CRS_DEP_TIME      int64
CARRIER          object
DEP_TIME          int64
DEST             object
DISTANCE          int64
FL_DATE          object
FL_NUM            int64
ORIGIN           object
Weather           int64
DAY_WEEK          int64
DAY_OF_MONTH      int64
TAIL_NUM         object
Flight Status    object
dtype: object

In [7]:
data.describe()

CRS_DEP_TIME     DEP_TIME     DISTANCE       FL_NUM      Weather  \
count   2201.000000  2201.000000  2201.000000  2201.000000  2201.000000   
mean    1371.938664  1369.298955   211.871422  3815.086324     0.014539   
std      432.697149   442.462754    13.316815  2409.750224     0.119725   
min      600.000000    10.000000   169.000000   746.000000     0.000000   
25%     1000.000000  1004.000000   213.000000  2156.000000     0.000000   
50%     1455.000000  1450.000000   214.000000  2385.000000     0.000000   
75%     1710.000000  1709.000000   214.000000  6155.000000     0.000000   
max     2130.000000  2330.000000   229.000000  7924.000000     1.000000   

          DAY_WEEK  DAY_OF_MONTH  
count  2201.000000   2201.000000  
mean      3.905498     16.024989  
std       1.903149      8.677390  
min       1.000000      1.000000  
25%       2.000000      8.000000  
50%       4.000000     16.000000  
75%       5.000000     23.000000  
max       7.000000     31.000000

Check missing values

In [9]:
data.isnull().sum()

CRS_DEP_TIME     0
CARRIER          0
DEP_TIME         0
DEST             0
DISTANCE         0
FL_DATE          0
FL_NUM           0
ORIGIN           0
Weather          0
DAY_WEEK         0
DAY_OF_MONTH     0
TAIL_NUM         0
Flight Status    0
dtype: int64

Dealing with Categorical Variables

features = ['DAY_WEEK', 'CRS_DEP_TIME', 'ORIGIN', 'DEST', 'CARRIER']

target = 'Flight Status'

For Naive Bayes, all the values should be in some categories.

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2201 entries, 0 to 2200
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   CRS_DEP_TIME   2201 non-null   int64 
 1   CARRIER        2201 non-null   object
 2   DEP_TIME       2201 non-null   int64 
 3   DEST           2201 non-null   object
 4   DISTANCE       2201 non-null   int64 
 5   FL_DATE        2201 non-null   object
 6   FL_NUM         2201 non-null   int64 
 7   ORIGIN         2201 non-null   object
 8   Weather        2201 non-null   int64 
 9   DAY_WEEK       2201 non-null   int64 
 10  DAY_OF_MONTH   2201 non-null   int64 
 11  TAIL_NUM       2201 non-null   object
 12  Flight Status  2201 non-null   object
dtypes: int64(7), object(6)
memory usage: 223.7+ KB


In [11]:
data.head()

CRS_DEP_TIME CARRIER  DEP_TIME DEST  DISTANCE     FL_DATE  FL_NUM ORIGIN  \
0          1455      OH      1455  JFK       184  01/01/2004    5935    BWI   
1          1640      DH      1640  JFK       213  01/01/2004    6155    DCA   
2          1245      DH      1245  LGA       229  01/01/2004    7208    IAD   
3          1715      DH      1709  LGA       229  01/01/2004    7215    IAD   
4          1039      DH      1035  LGA       229  01/01/2004    7792    IAD   

   Weather  DAY_WEEK  DAY_OF_MONTH TAIL_NUM Flight Status  
0        0         4             1   N940CA        ontime  
1        0         4             1   N405FJ        ontime  
2        0         4             1   N695BR        ontime  
3        0         4             1   N662BR        ontime  
4        0         4             1   N698BR        ontime

In [12]:
#changing DAY_WEEK to categorical variables

data['DAY_WEEK_cat'] = data['DAY_WEEK'].astype('category')
data['DAY_WEEK_cat'].value_counts()

5    391
4    372
3    320
1    308
2    307
7    253
6    250
Name: DAY_WEEK_cat, dtype: int64

In [13]:
#create hourly bins departure time

data['CRS_DEP_TIME_cat'] = [round(t/500) for t in data['CRS_DEP_TIME']]
data['CRS_DEP_TIME_cat'] = data['CRS_DEP_TIME_cat'].astype('category')
data['CRS_DEP_TIME_cat'].value_counts()

3    966
2    582
4    394
1    259
Name: CRS_DEP_TIME_cat, dtype: int64

In [14]:
#changing Flight Status to categorical variables

data['Flight Status'] = data['Flight Status'].astype('category')
data['Flight Status'].value_counts()

ontime     1773
delayed     428
Name: Flight Status, dtype: int64

In [15]:
features = ['DAY_WEEK_cat', 'CRS_DEP_TIME_cat', 'ORIGIN', 'DEST', 'CARRIER']

target = 'Flight Status'

In [16]:
X = pd.get_dummies(data[features])
y = data['Flight Status'].astype('category')

classes = list(y.cat.categories)
print(classes)

['delayed', 'ontime']


Train Test Split

In [17]:
from sklearn.model_selection import train_test_split as tts

X_train, X_valid, y_train, y_valid = tts(X, y, train_size=0.7)

In [18]:
print(y_train.value_counts())

ontime     1237
delayed     303
Name: Flight Status, dtype: int64


In [19]:
print(y_valid.value_counts())

#imbalance is present, training ontime is 4 times delayed, validation ontime is 4.3 times delayed

ontime     536
delayed    125
Name: Flight Status, dtype: int64


In [20]:
#to make it balanced use stratify = y

from sklearn.model_selection import train_test_split as tts

X_train, X_valid, y_train, y_valid = tts(X, y, train_size=0.7, random_state=1, stratify=y)

In [21]:
print(y_train.value_counts())

ontime     1241
delayed     299
Name: Flight Status, dtype: int64


In [22]:
print(y_valid.value_counts())

#both ontime is 4.1 times delayed, so balanced

ontime     532
delayed    129
Name: Flight Status, dtype: int64


Running Algorithm

In [23]:
from sklearn.naive_bayes import MultinomialNB as MNB
model = MNB()

In [24]:
model.fit(X_train, y_train)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [25]:
#make predictions on validation set

y_pred = model.predict(X_valid)

In [26]:
results = pd.DataFrame({'actual':y_valid, 'predicted':model.predict(X_valid)})
results.head()

actual predicted
1751   ontime    ontime
96    delayed    ontime
1899   ontime    ontime
681    ontime    ontime
1258  delayed    ontime

In [27]:
from sklearn import metrics

In [28]:
print(metrics.accuracy_score(results.actual, results.predicted))

0.8018154311649016


In [29]:
print(metrics.confusion_matrix(results.actual, results.predicted))

#3 are actual values True and predicted values also True
#527 are actual values False and predicted values also False

[[  3 126]
 [  5 527]]


In [30]:
print(metrics.classification_report(results.actual, results.predicted))

              precision    recall  f1-score   support

     delayed       0.38      0.02      0.04       129
      ontime       0.81      0.99      0.89       532

    accuracy                           0.80       661
   macro avg       0.59      0.51      0.47       661
weighted avg       0.72      0.80      0.72       661



So we have predicted using Multinomial NB an accuracy of 80.18% whether the flights got delayed or were ontime. 